crates/assembly/src/mast_forest_builder.rs

```diff
@@ -489,27 +544,40 @@ impl MastForestBuilder {
     ///
     /// If other decorators are already present, the new decorators are added to the end of the
     /// list.
-    pub fn append_before_enter(&mut self, node_id: MastNodeId, decorator_ids: &[DecoratorId]) {
-        self.mast_forest[node_id].append_before_enter(decorator_ids);
+    pub fn append_before_enter(
+        &mut self,
+        node_id: MastNodeId,
+        decorator_ids: Vec<DecoratorId>,
+    ) -> Result<(), MastForestError> {
+        // Extract the existing node and convert it to a builder
+        let mut decorated_builder = self.mast_forest[node_id].clone().to_builder();
+        decorated_builder.append_before_enter(decorator_ids);
+        let new_node_fingerprint =
+            decorated_builder.fingerprint_for_node(&self.mast_forest, &self.hash_by_node_id)?;
+        self.mast_forest[node_id] = decorated_builder.build(&self.mast_forest)?;
```

Comment:
Don't we also need to update `self.node_id_by_fingerprint` with the new mapping (and possibly remove the previous/stale one)?

Reply:
Done!